rfcs/0005-kinesis-proxy-json.md

```diff
@@ -0,0 +1,320 @@
+- Feature Name: kinesis-proxy-json
+- Start Date: 2021-06-02
+- RFC PR: mbta/technology-docs#5
+- Asana task: [Secure connection between socket_proxy and rtr](https://app.asana.com/0/881264583703207/1200290609745412/f)
+- Status: Proposed
+
+# Summary
+
+Send OCS messages as JSON-formatted CloudEvents to AWS Kinesis
+
+# Motivation
+
+Currently, SocketProxy proxies OCS messages to each RTR instance running in
+AWS. This method of sending Occupancy Control System (OCS) messages has
+several issues we are trying to address:
+
+- insecure: the messages travel unencrypted over the public internet
+- insecure: anyone from the MBTA network can forge OCS messages
+- unreliable: if an RTR instance isn't running (crashed, or during a
+  restart/redeploy) when a message comes in, that message is lost
+- unreliable: if the production RTR instance isn't running, then the log of
+  messages written to S3 (and used by OPMI) will not have those messages
+- wasteful: we send each message four separate times to four RTR instances
+- missing metadata: the messages do not include the date they were sent, or
+  the time the messages were received by SocketProxy 
+
+To address these issues, we'll create KinesisProxy, a new Elixir application,
```

Comment:
What about a more descriptive name? I'm thinking something like `OCSMessageProxy` or something to that effect. Kinesis is something of an implementation detail, and I think it would be helpful to 1) have the windows service clearly map to the underlying elixir app, and 2) have that windows service's purpose be clear.

It's also not really a generic KinesisProxy in the same way that SocketProxy agnostically proxies TCP sockets. It at the very least looks at the stream to parse at EOT and sends a Kinesis message called `ocs_message`. I thought about actually making it a generic KinesisProxy (and open sourced and usable by all), and having the specific OCS behavior swappable by the client, but that seems a bit overkill for our purposes here. Plus, we're wanting this app over time to get _more_ OCS specific and start parsing the messages and uploading different types.

Reply:
Those are all good points! I've renamed it to `Tricycle` for now, and mentioned that the exact name can be resolved outside the RFC :)